clang/include/clang/Serialization/ModuleFile.h

```diff
@@ -288,10 +288,12 @@ class ModuleFile {
   /// for the entry is SLocEntryOffsetsBase + SLocEntryOffsets[i].
   uint64_t SLocEntryOffsetsBase = 0;
 
-  /// Offsets for all of the source location entries in the
-  /// AST file.
+  /// Stream bit offsets for all of the source location entries in the AST file.
   const uint32_t *SLocEntryOffsets = nullptr;
 
+  /// SLocEntry offsets that have been loaded from the AST file.
+  std::vector<SourceLocation::UIntTy> SLocEntryOffsetLoaded;
```

Comment:
I assume this is relatively small memory use in practice?

Reply:
This ends up being ~10MB for single TU using 37 modules from the Darwin SDK. I'm not thrilled about that, what do you think? We could store these offsets in `SourceManager::LoadedSLocEntryTable` and mark the fact that the entry has not been fully deserialized in a counterpart to `llvm::BitVector SLocEntryLoaded`.

Reply:
I don't feel that strongly about 10 MB in a large TU; your suggestion seems reasonable, but could be a follow up.